File: scripts/generate_projects_manifest.py
# Ejecuta: python3 scripts/generate_projects_manifest.py
import os, json, sys

ROOT = os.getcwd()
PROJ_DIR = os.path.join(ROOT, "proyectos")
OUT_FILE = os.path.join(ROOT, "projects.json")

def humanize(name):
    out = name.replace("-", " ").replace("_", " ")
    return " ".join([w.capitalize() for w in out.split()])

def list_html_entries(dir_path, base=""):
    items = []
    for name in os.listdir(dir_path):
        if name.startswith("."):
            continue
        full = os.path.join(dir_path, name)
        rel  = f"{base}/{name}" if base else name

        if os.path.isdir(full):
            index_path = os.path.join(full, "index.html")
            if os.path.isfile(index_path):
                items.append({"title": humanize(name), "path": f"{rel}/index.html", "__isIndex": True})
            items.extend(list_html_entries(full, rel))
        elif os.path.isfile(full) and name.lower().endswith(".html"):
            if name.lower() == "index.html":
                continue
            items.append({"title": humanize(name[:-5]), "path": rel})
    return items

def main():
    if not os.path.isdir(PROJ_DIR):
        print("No existe /proyectos")
        sys.exit(1)

    grupos = []
    for name in os.listdir(PROJ_DIR):
        path = os.path.join(PROJ_DIR, name)
        if not os.path.isdir(path) or name.startswith("."):
            continue
        items = list_html_entries(path, name)
        if items:
            grupos.append({"nombre": name, "items": items})

    manifest = {"grupos": grupos}
    with open(OUT_FILE, "w", encoding="utf-8") as f:
        json.dump(manifest, f, ensure_ascii=False, indent=2)
    print(f"✅ Manifest generado: {OUT_FILE}")

if __name__ == "__main__":
    main()

File: scripts/generate-projects-manifest.mjs
// Ejecuta: node scripts/generate-projects-manifest.mjs
import { readdir, stat, writeFile } from "node:fs/promises";
import { join } from "node:path";

const ROOT = process.cwd();
const PROJ_DIR = join(ROOT, "proyectos");
const OUT_FILE = join(ROOT, "projects.json");

function humanize(name){
  return name.replace(/[-_]/g," ").replace(/\b\w/g, m=>m.toUpperCase());
}

async function listHtmlEntries(dir, base=""){
  const entries = await readdir(dir, { withFileTypes:true });
  const items = [];
  for (const ent of entries){
    if (ent.name.startsWith(".")) continue;
    const full = join(dir, ent.name);
    const rel  = base ? `${base}/${ent.name}` : ent.name;

    if (ent.isDirectory()){
      // Si tiene index.html, añadir entrada
      try{
        const st = await stat(join(full,"index.html"));
        if(st.isFile()){
          items.push({ title: humanize(ent.name), path: `${rel}/index.html`, __isIndex: true });
        }
      }catch{}
      // Recursión
      const sub = await listHtmlEntries(full, rel);
      sub.forEach(it=>items.push(it));
    } else if (ent.isFile() && ent.name.toLowerCase().endsWith(".html")){
      if (ent.name.toLowerCase() === "index.html") continue;
      items.push({ title: humanize(ent.name.replace(/\.html$/i,"")), path: rel });
    }
  }
  return items;
}

(async ()=>{
  try{
    const grupos = [];
    const top = await readdir(PROJ_DIR, { withFileTypes:true });
    for (const ent of top){
      if (!ent.isDirectory() || ent.name.startsWith(".")) continue;
      const groupDir = join(PROJ_DIR, ent.name);
      const items = await listHtmlEntries(groupDir, ent.name);
      if (items.length) grupos.push({ nombre: ent.name, items });
    }
    await writeFile(OUT_FILE, JSON.stringify({ grupos }, null, 2), "utf8");
    console.log(`✅ Manifest generado: ${OUT_FILE}`);
  }catch(err){
    console.error("❌ Error generando manifest:", err);
    process.exit(1);
  }
})();
